### ISSUE TITLE
Interface matching rules

### ISSUE THREAD

**Issue by litherum** [gpuweb/gpuweb#644]
We need a section on the specific interface matching rules of WGSL.

**Comment by kvark** [comment]
A short description of logic I proposed is the following. We could see the interface rules applied between consequent shader stages as well as between the fixed-function vertex input and the vertex shader, and between the fragment shader and the fixed function pixel export.

For each interface, it's considered valid as long as the set of inputs consumed by the next stage is a subset of the set of outputs produced by the last stage. Examples:
  - some vertex attributes/buffers may be unused by the VS
  - some varying outputs of VS may be unused by the FS
  - some fragment outputs of the FS may be ignored by the pipeline

**Comment by dneto0** [comment]
In the meeting of [2020-06-16](https://docs.google.com/document/d/11RW9hpfzSvZYpLGBm3uhKwoXDvJI42N7jWP_nOB-k2E/edit#heading=h.hp3f2zbslxr9) I mentioned that I needed to do some homework to see whether Vulkan shader builtin variables affect intra-pipeline shader interface shader matching rules.

I've reviewed Vulkan's shader matching rules ([section 1.14.1.3](https://www.khronos.org/registry/vulkan/specs/1.2/html/vkspec.html#interfaces-iointerfaces-matching)) and I've concluded that builtin-variables do *not* affect interface matching, for the features enabled in WebGPU.    Basically, builtin variables only affect the intra-pipeline shader interface when there is more than one non-Fragment shader.  WebGPU only has vertex+fragment and compute pipelines.

**Comment by kvark** [comment]
Sorry to bring this up, but things like `gl_FragDepth` are also going to be built-ins, right? Do you suggest we ignore this for validation purposes, even if the pipeline doesn't have a depth attachment?

**Comment by kvark** [comment]
Also, I see an interesting requirement of MSL mentioned here - https://github.com/KhronosGroup/MoltenVK/pull/931
We need to adjust our rules for that, I don't want shader translation of one stage to know about other stages (thus, I don't want us to try working around this in MSL).

**Comment by dj2** [comment]
What do you mean by having `gl_FragDepth` as a builtin, that it would work like GLSL where it just exists in the shader? There are currently no plans to do that in WGSL. You have to  declare everything you're using.

e.g. `[[builtin position]] var<out> gl_Position : vec4<f32>;` happens to use the `gl_` naming, but that was just for ease of porting from the original GLSL source. There is no restriction on what that name could be.

**Comment by johnkslang** [comment]
`gl_FragData` (which I assume comes from the `[[ ... ]]`, not the spelling) used to have pretty special abilities, like broadcast, but that was dropped for SPIR-V.  I assume WebGPU will also drop it, and just have outputs at numbered locations, which is seems a validator or pipeline creator would verify line up with the attachments.

**Comment by kvark** [comment]
@dj2 are we going to have an ability for the fragment shader to output depth directly? If yes, how do we define the interface here? In GLSL there is no "location = " for depth output, it's just a `gl_FragDepth` builtin. So the problem can be stated as: this is builtin that affects the interface, contrary to what @dneto0 stated:
> I've concluded that builtin-variables do not affect interface matching

**Comment by dj2** [comment]
Couldn't you do something like `[[builtin frag_depth, location 2]] var<out> my_fragdepth : f32` ?

**Comment by magcius** [comment]
> We need to adjust our rules for that, I don't want shader translation of one stage to know about other stages

Why not? WebGPU has PSOs, so all shader translation can be done at PSO creation time if necessary, and even full-pipeline optimization can be attempted. It should be expected that PSO creation is somewhat expensive.

**Comment by kvark** [comment]
@dj2 
> Couldn't you do something like [[builtin frag_depth, location 2]] var<out> my_fragdepth : f32 ?

You could, but that would look rather awkward, since there is nothing to refer to the "location 2" on the API side.

@magcius doing things early is the landmark of todays explicit graphics API. We don't want to delay something to later just because we can. Shader translation may be not the most expensive part of the pipeline creation, but it's still work:
  - parsing
  - type checking
  - uniformity/control flow analysis
  - validating

Moreover, if user is creating 1000 pipelines using the same fragment shader, it would be much beneficial if the work for this fragment shader was naturally reused as a part of the `GPUShaderModule` building.

All in all, implementations are free to defer shader translation as well as validation, we are totally cool with that. We are not cool with the idea that this will be forced upon. We'd like to translate the shader at `createShaderModule` and know about their interface at that point.

As for the full-pipeline optimization, it's not clear if this is something we are going to do anyway. It might be best to leave this to the drivers.

**Comment by dj2** [comment]
Ok, I don't think I know enough about how `gl_FragDepth` works. If I want to access it on the API side, it's ... just there? How does WebGPU know to read it from `location 2` if that's what I said in my shader, is it by introspection or something? Or is it because of the`built_in frag_depth` and, if it matches up that way, does it have a location at all, or how does the location come into it?

**Comment by kvark** [comment]
@dj2 it has an implicit location, since there can be only one depth attachment in a render pass. It's basically "just there".
What I'm trying to say in this thread is that the use of `gl_FragDepth` (or whatever we call it in WGSL) should be considered a part of the interface.

**Comment by dj2** [comment]
Are there other built-ins with this same implicit location behaviour or is `frag_depth` the only one?

**Comment by Kangz** [comment]
Maybe `gl_SampleMask`, `gl_SecondaryFragColor` (when we get dual-source blending) and maybe more.

**Comment by johnkslang** [comment]
Note that some people (e.g., @dneto0) was answering for "intra" interfaces, like between stages.  The input to the first (vertex) stage and the output from the last (fragment) stage are a different kind of interface.  

Those don't participate in "do these two compilation units have matching interfaces?" and instead participate in "does the set of shader stages interface correctly to the rest of the pipeline?".  For that you need to line up:

1. inputs streamed into the vertex stage
2. buffers read/written by all stages
3. outputs streamed out from the fragment stage

**Comment by magcius** [comment]
@kvark you can do parsing, validation, etc. at shader creation time, and defer MSL codegen. Presumably, you could keep a cache of inputs to "IRs" internally, and then a second cache for when the VS output layout is known.

Anyway, if we want to require complete location compatibility between VS/FS stages, that's fine with me too. Typically that's just handled at PSO creation time in the drivers, anyway.

**Comment by magcius** [comment]
@Kangz 

`gl_SecondaryFragColor` doesn't exist in an MRT world. DSB is a separate (explicit) output for SPIR-V / GLSL:

```
layout(location = 0, index = 0) out vec4 col1;
layout(location = 0, index = 1) out vec4 col2;
```

`gl_SampleMask` and SPV_EXT_shader_stencil_export's StencilRef appear to be builtins the same way. But this is unrelated to VS/FS stage compatibility -- not sure how it's relevant here.

**Comment by kvark** [comment]
>  But this is unrelated to VS/FS stage compatibility -- not sure how it's relevant here.

There is clearly a miscommunication on what "interface" is within this issue.
I read it initially as - interface between the pipeline and each of the shader stages, as well as between shader stages. Sounds like @dneto0 and @johnkslang don't consider the interface with the pipeline here. Is there a value in looking at the shader stages compatibility in isolation?

**Comment by johnkslang** [comment]
I'm not trying to discount any particular interface, just saying that if the question is about "WGSL interface matching rules" (see OP) then I understand thinking it's between compilation units and/or stages written in WGSL.

Maybe all interfaces were meant, but that's really not just as WGSL subject, its more of a pipeline subject or WGSL/pipeline/API/app subject.

**Comment by magcius** [comment]
Sorry, this I misread the thread and thought @dj2's recent comments about gl_FragDepth were because of the Metal issue that was brought up about matching VS/FS compatibility, and was confused how they were related. Now I see they were a continuation of a previous discussion.

**Comment by grorg** [comment]
Discussed at the [2020-06-23 Virtual F2F meeting](https://docs.google.com/document/d/1WI9PzwXy39fFHuZLUnEQtSplwfmFaZ2oVb0gXJsA2P4/)

**Comment by kvark** [comment]
Tentative agreement on the call:
  - varyings between stage are matched by user-specified integer locations
  - outputs from the previous stage are either not used in the next stage
  - or they have to have exactly the same type and the interpolation qualifier

There is a research direction on how we can get closer to the final DXIL (on D3D12) at `createShaderModule` time. One issue is the dynamic-offset buffers, which need to have the code patched, and we currently don't know about them till the pipeline creation time. We may consider having the dynamic/static offset buffer being distinct in WGSL to alleviate this.

More investigation to do here with Microsoft's help.

**Comment by dneto0** [comment]
We also decided to not support the equivalent of the Component attribute from Vulkan.  Only use the Location component.

**Comment by dneto0** [comment]
Sorry for the confusion above.  I meant that for checking compatibility **between vertex shader and fragment shader**, the builtins don't affect the rules.  @johnkslang hopefully cleared that up.

Interface matching for the rest of the pipeline state is another matter entirely.

**Comment by grorg** [comment]
Discussed at the [2020-07-07 meeting](https://docs.google.com/document/d/1jiRiUAEBkG1UPFZTTeEx81HCXlrTeucCrfATIFBhzzI/).

**Comment by dneto0** [comment]
From the last meeting I had homework to check on the Vulkan rules about location assignment for input/output interfaces vertex and fragment shaders.  See [14 Shader interfaces](https://www.khronos.org/registry/vulkan/specs/1.2/html/vkspec.html#interfaces) in the Vulkan 1.2 spec for full details.

User data must map to Location and Component.  We already discussed that Metal doesn't have Component, so let's ignore those.  Each Location can fit up to a `vec4<f32>` worth of data (16 bytes, or 4 components)
Locations are specified in the shader as literal numbers, from 0 to a maximum number of components.  The mandated minimum max is 64, which translates to 16 Locations.

Consequences:

Location assignments, along with type compatibility, is a key element to determining shader compatibility within a pipeline.  It is important that location assignment remain under user control, and so they must be explicit.
I think it's wrong to have automatic remapping by the WebGPU implementation.

**Comment by pow2clk** [comment]
We owe the WGSL group the D3D perspective on this. We have a lot of specs to go over and a lot of other demands on our time. We can commit to having the information by the meeting on the 28th of July. I hope that's acceptable.

**Comment by Kangz** [comment]
There were at least two open questions after the 2020-07-07 call:

 - Should the interface match exactly between FS and VS (it is D3D constraint, but what's the cost of doing it at pipeline creation?)
 - Should the locations being bounded between FS and VS or should they be arbitrary?

Another question that didn't seem answered in that meeting is whether it is allowed for FS to produce a superset of outputs compared to the render passes' attachment or if they should match exactly. (we already know that having VS consume a subset of the attributes is ok).

**Comment by Kangz** [comment]
All of these questions have been resolved in the spec, with #2060 and #2038 for two follow-up questions. Closing this.
